From ea3f5e334fbee20fcc7568e1ba42ce4538b54a36 Mon Sep 17 00:00:00 2001
From: Paul Keith <javelinanddart@gmail.com>
Date: Thu, 2 Aug 2018 15:27:50 +0200
Subject: [PATCH 7/9] sdk: Allow controllable brightness for non-RGB segmented
 battery LEDs

* For non-RGB segmented battery LEDs, we currently don't allow any kind
  of brightness control, because the alpha channel is taken up by the level
* To remedy this, set the brightness in the color channel like we do for
  setting panel backlight brightness by setting the brightness as R/G/B
* This will require any devices with RGB segmeneted battery LEDs to not
  set the LIGHTS_ADJUSTABLE_BATTERY_LED_BRIGHTNESS flag, and non-RGB
  devices to set it in order to allow control

Change-Id: I4e47861643e0b2b8766af0f2ff275069fc580108
---
 .../notification/LineageBatteryLights.java    | 26 +++++++++++++++----
 1 file changed, 21 insertions(+), 5 deletions(-)

diff --git a/sdk/src/java/org/lineageos/internal/notification/LineageBatteryLights.java b/sdk/src/java/org/lineageos/internal/notification/LineageBatteryLights.java
index 121ac8e..913389f 100644
--- a/sdk/src/java/org/lineageos/internal/notification/LineageBatteryLights.java
+++ b/sdk/src/java/org/lineageos/internal/notification/LineageBatteryLights.java
@@ -180,11 +180,27 @@ public final class LineageBatteryLights {
         if (ledValues.getColor() != 0) {
             ledValues.setEnabled(true);
         }
-        // If lights HAL does not support adjustable brightness then
-        // scale color value here instead.
-        if (mCanAdjustBrightness && !mHALAdjustableBrightness) {
-            ledValues.applyAlphaToBrightness();
-            ledValues.applyBrightnessToColor();
+        if (mCanAdjustBrightness) {
+            if (!mHALAdjustableBrightness) {
+                // If lights HAL does not support adjustable brightness then
+                // scale color value here instead.
+                ledValues.applyAlphaToBrightness();
+                ledValues.applyBrightnessToColor();
+            } else if (mUseSegmentedBatteryLed) {
+                // For non-RGB segmented LEDs, we must set the brightness
+                // as the color, since the alpha channel contains the level
+                int color = 0;
+                if (mZenMode == Global.ZEN_MODE_OFF) {
+                   color |= mBatteryBrightnessLevel;
+                   color |= mBatteryBrightnessLevel << 8;
+                   color |= mBatteryBrightnessLevel << 16;
+                } else {
+                   color |= mBatteryBrightnessZenLevel;
+                   color |= mBatteryBrightnessZenLevel << 8;
+                   color |= mBatteryBrightnessZenLevel << 16;
+                }
+                ledValues.setColor(color);
+            }
         }
         // If LED is segmented, reset brightness field to battery level
         // (applyBrightnessToColor() changes it to 255)
-- 
2.17.1

